src/compiler/emitter.ts

```diff
@@ -1529,13 +1532,26 @@ var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, ge
                         }
                     }
 
-                    if (languageVersion !== ScriptTarget.ES6) {
+                    if (languageVersion < ScriptTarget.ES6) {
                         const declaration = resolver.getReferencedDeclarationWithCollidingName(node);
                         if (declaration) {
                             write(getGeneratedNameForNode(declaration.name));
                             return;
                         }
                     }
+                    else if (resolver.getNodeCheckFlags(node) & NodeCheckFlags.BodyScopedClassBinding) {
+                        // Due to the emit for class decorators, any reference to the class from inside of the class body
+                        // must instead be rewritten to point to a temporary variable to avoid issues with the double-bind
+                        // behavior of class names in ES6.
+                        const declaration = resolver.getReferencedValueDeclaration(node);
+                        if (declaration && declaration.kind === SyntaxKind.ClassDeclaration) {

```

Comment:
when declaration kind might not be `ClassDeclaration`? I think that if checker has set `BodyScopedClassBinding` flag it should be enough to state that referenced declaration is class